Makefile

```diff
@@ -126,34 +131,75 @@ endef
-
-update-gotestsum:
-	@printf $(COLOR) "Install/update gotestsum..."
-	@go install gotest.tools/gotestsum@v1.11
-
-update-proto-plugins:
-	@printf $(COLOR) "Install/update proto plugins..."
-	@go install google.golang.org/protobuf/cmd/protoc-gen-go@latest
-	@go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest
-	@go install -modfile build/go.mod go.temporal.io/api/cmd/protoc-gen-go-helpers
-	@go install -modfile build/go.mod go.temporal.io/api/cmd/protogen
-
-update-proto-linters:
-	@printf $(COLOR) "Install/update proto linters..."
-	@go install github.com/googleapis/api-linter/cmd/api-linter@v1.32.3
-	@go install github.com/bufbuild/buf/cmd/buf@v1.6.0
+$(STAMPDIR):
+	@mkdir -p $(STAMPDIR)
+
+$(LOCALBIN):
+	@mkdir -p $(LOCALBIN)
+
+# When updating the version, update the golangci-lint GHA workflow as well.
+.PHONY: golangci-lint
+GOLANGCI_LINT_VERSION := v1.57.2
+GOLANGCI_LINT := $(LOCALBIN)/golangci-lint-$(GOLANGCI_LINT_VERSION)
+$(GOLANGCI_LINT): $(LOCALBIN)
+	$(call go-install-tool,$(GOLANGCI_LINT),github.com/golangci/golangci-lint/cmd/golangci-lint,$(GOLANGCI_LINT_VERSION))
+
+GOTESTSUM_VER := v1.11
+GOTESTSUM := $(LOCALBIN)/gotestsum-$(GOTESTSUM_VER)
+$(GOTESTSUM): | $(LOCALBIN)
+	$(call go-install-tool,$(GOTESTSUM),gotest.tools/gotestsum,$(GOTESTSUM_VER))
+
+API_LINTER_VER := v1.32.3
+API_LINTER := $(LOCALBIN)/api-linter-$(API_LINTER_VER)
+$(API_LINTER): | $(LOCALBIN)
+	$(call go-install-tool,$(API_LINTER),github.com/googleapis/api-linter/cmd/api-linter,$(API_LINTER_VER))
+
+BUF_VER := v1.6.0
+BUF := $(LOCALBIN)/buf-$(BUF_VER)
+$(BUF): | $(LOCALBIN)
+	$(call go-install-tool,$(BUF),github.com/bufbuild/buf/cmd/buf,$(BUF_VER))
+
+GO_API_VER := v1.29.0
+PROTOGEN := $(LOCALBIN)/protogen-$(GO_API_VER)
+$(PROTOGEN): | $(LOCALBIN)
+	$(call go-install-tool,$(PROTOGEN),go.temporal.io/api/cmd/protogen,$(GO_API_VER))
+
+
+# The following tools need to have a consistent name, so we use a versioned stamp file to ensure the version we want is installed
+# while installing to an unversioned binary name.
+GOIMPORTS_VER := v0.20.0
+GOIMPORTS := $(LOCALBIN)/goimports
+$(STAMPDIR)/goimports-$(GOIMPORTS_VER): | $(STAMPDIR) $(LOCALBIN)
+	$(call go-install-tool,$(GOIMPORTS),golang.org/x/tools/cmd/goimports,$(GOIMPORTS_VER))
+	@touch $@
+$(GOIMPORTS): $(STAMPDIR)/goimports-$(GOIMPORTS_VER)
+
+# Mockgen is called by name throughout the codebase, so we need to keep the binary name consistent
+MOCKGEN_VER := v1.7.0-rc.1
+MOCKGEN := $(LOCALBIN)/mockgen
+$(STAMPDIR)/mockgen-$(MOCKGEN_VER): | $(STAMPDIR) $(LOCALBIN)
+	$(call go-install-tool,$(MOCKGEN),github.com/golang/mock/mockgen,$(MOCKGEN_VER))
+	@touch $@
+$(MOCKGEN): $(STAMPDIR)/mockgen-$(MOCKGEN_VER)
+PROTOC_GEN_GO_VER := v1.33.0
+PROTOC_GEN_GO := $(LOCALBIN)/protoc-gen-go
+$(STAMPDIR)/protoc-gen-go-$(PROTOC_GEN_GO_VER): | $(STAMPDIR) $(LOCALBIN)
+	$(call go-install-tool,$(PROTOC_GEN_GO),google.golang.org/protobuf/cmd/protoc-gen-go,$(PROTOC_GEN_GO_VER))
+	@touch $@
```

Comment:
So if we install protoc v1.33 and touch `.stamp/protoc-v1.33`, and then install v1.40 and touch `.stamp/protoc-v1.40`, and then go to an old release branch that's on v1.33, it will leave v1.40 installed, right? It seems like we need not just the presence of a bunch of files, but a single file that contains the version and compare to that. Or else do something like `rm .stamp/protoc-*` (but that only works if no tool name is a prefix of another tool name, which actually they are)

Reply:
If we kept a subdirectory for each tool we could do that, e.g. `.stamp/protoc/vFOO`. That may be the best approach.

This is a problem with using unversioned binaries, but protoc (for example) needs that. I'll play around with it once I'm back next week